gobase64.go

```diff
@@ -1,161 +1,139 @@
 package main
 
 import (
+	"container/list"
+	"encoding/binary"
 	"errors"
 	"io"
 	"log"
+	"math/bits"
 	"os"
 )
 
-func main() {
-	reader := os.Stdin
-	writer := os.Stdout
+// our worker count
+const WORKERS = 4
 
-	inbuf := make([]byte, 1024)
-	outbuf := make([]byte, 1368)
+type B64Chars struct {
+	value []byte
+}
 
-	for {
-		octetsIn, err := reader.Read(inbuf)
-		if err == io.EOF {
-			break
-		} else if err != nil {
-			log.Fatal("Unexpected IO error", err)
+// Not immutable but runtime array slicing to create a pseudo-immutable array is slower
+var charset = [64]byte{
+	'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R',
+	'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z', 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j',
+	'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z', '0', '1',
+	'2', '3', '4', '5', '6', '7', '8', '9', '+', '/'}
+
+// sextuple to chars
+// assumes output is initially empty
+func getChars(buffer *[]byte, padding int) ([]byte, error) {
+	if padding > 2 {
+		return []byte{}, nil
+	}
+	var out []byte
+	var in = *buffer
+	var l = len(in)
+	if l != 3 {
+		return nil, errors.New("expected 3 bytes")
+	}
+	if padding > 0 {
+		in[2] = 0x00
+	}
+	if padding > 1 {
+		in[1] = 0x00
+	}
+	if l > 0 {
+		var unsigned = binary.BigEndian.Uint32(append([]byte{0x00}, in...))
+		out = append(out, charset[bits.RotateLeft32((unsigned&0xFC0000), -18)])
+		out = append(out, charset[bits.RotateLeft32((unsigned&0x3F000), -12)])
+		if padding < 2 {
+			out = append(out, charset[bits.RotateLeft32((unsigned&0x0FC0), -6)])
 		}
-
-		n, err := Encode(&inbuf, octetsIn, &outbuf)
-
-		if err != nil {
-			log.Fatal("Unexpected Base64 encoding error", err)
+		if padding < 1 {
+			out = append(out, charset[(unsigned&0x003F)])
 		}
-
-		writer.Write(outbuf[:n])
 	}
+	return out, nil
 }
 
-func Encode(input *[]byte, octets int, output *[]byte) (int, error) {
-	triplets := octets / 3
-	remainder := octets % 3
-	var outlen int
-	if remainder == 0 {
-		outlen = triplets * 4
-	} else {
-		outlen = triplets*4 + 4
-	}
+func btoa_parallel(reader io.Reader, writer io.Writer) {
+	c := make(chan *B64Chars, 16)
 
-	if outlen > len(*output) {
-		return 0, errors.New("output buffer is too small")
+	// load 3^10 bytes
+	var inbuf = make([]byte, 3)
+	var bl = 0
+	outbuffer := list.New()
+	outbuffer.Init()
+	for {
+		r, rerr := reader.Read(inbuf)
+		bl += r
+		if rerr != nil && rerr != io.EOF {
+			log.Fatal("Unexpected IO error", rerr)
+		}
+		// offload the getChars to a byte array
+		go func() {
+			s, err := getChars(&inbuf, 3-r)
+			if err != nil {
+				log.Fatal("Unexpected Base64 encoding error", err)
+			}
+			chars := new(B64Chars)
+			chars.value = s
+			c <- chars
+		}()
+		outbuffer.PushBack(<-c)
```

Comment:
This blocks, is that right? Will the main thread read the input then wait for the b64 thread to encode it? I'm thinking that we'd want to send each chunk along with its sequence number to an input channel, then have the main thread block on an output channel and do resequencing of outputs so that many threads can run at the same time. 

I'll have to look up performance testing tools in go later. We're getting to that point. :D